console: decouple count query from fetch data query

beerose opened the pull request:
### Description

This PR decouples count query from fetch data query. 

### Changelog

- [x] `CHANGELOG.md` is updated with user-facing content relevant to this PR.

### Affected components

- [x] Console

### Related Issues
https://github.com/hasura/graphql-engine/issues/3793

beerose commented:
@rikinsk It should be fixed now.

wawhal commented on console/src/components/Services/Data/TableBrowseRows/ViewActions.js:
```diff
@@ -113,6 +101,50 @@ const vMakeRequest = () => {
+    const originalTable = getState().tables.currentTable;
+    dispatch({ type: V_COUNT_REQUEST_START });
+
+    const requestBody = {
```
Move to v1QueryUtils.

wawhal commented on console/src/components/Services/Data/TableBrowseRows/ViewActions.js:
```diff
@@ -57,29 +61,14 @@ const vMakeRequest = () => {
     dispatch({ type: V_REQUEST_PROGRESS, data: true });
 
     const requestBody = {
```
Move to v1QueryUtils as `generateSelectQuery = (tableDef, where, limit, offset, orderBy) => { ...`

wawhal commented on console/src/components/Services/Data/TableBrowseRows/ViewActions.js:
```diff
@@ -113,6 +101,50 @@ const vMakeRequest = () => {
+    };
+    return dispatch(requestAction(url, options)).then(
+      data => {
+        if (originalTable === getState().tables.currentTable) {
```
What is the reason behind this check?
Isn't `originalTable` defined as `getState().tables.currentTable` in the same scope?

beerose replied:
It's prev logic.

beerose replied:
The idea is that user may change table while request is pending. And there might be race condition with another fetch data query.

rikinsk replied:
This was buggy indeed. Fixed it now

wawhal commented on console/src/components/Services/Data/TableBrowseRows/ViewActions.js:
```diff
@@ -113,6 +101,50 @@ const vMakeRequest = () => {
+  };
+};
+
 const fetchManualTriggers = tableName => {
```
Move the query to v1QueryUtils.

wawhal commented on console/src/components/Services/Data/TableBrowseRows/ViewRows.js:
```diff
@@ -763,9 +764,7 @@ const ViewRows = ({
           <b className={styles.padd_small_right}>Selected:</b>
           {selectedRows.length}
           <button
```
Use the common Button component.

beerose replied:
Isn't it duplicated work, since we'll be soon integrating buttons from the design system?

wawhal commented on console/src/components/Common/TableCommon/DragFoldTable.js:
```diff
@@ -3,7 +3,10 @@
+import Spinner from '../Spinner/Spinner';
+import styles from '../Common.scss';
 
 Object.assign(ReactTableDefaults, {
```
Sorry I missed this in the `browse rows improvements` PR, but can you remove these defaults and pass them to the ReactTable component directly where it's used?

Just to make the code declarative. There's no way anybody will read these defaults if they are reading the component code. Do you agree?

beerose replied:
I removed it because:
1) defaultPageSize is already 10,
2) minRows wasn't enforced in react-table anyway.

wawhal commented on console/src/components/Common/TableCommon/DragFoldTable.js:
```diff
@@ -87,7 +90,11 @@ class DragFoldTable extends Component {
             })
           }
           folded={this.state.folded}
+          showPagination={!this.props.countLoading}
```
Is this needed?

beerose replied:
When the count is still loading there's no pagination displayed. Instead, there's a spinner.

rikinsk commented:
Also noticed one other thing. The bulk selected rowws should be reset when we change pages.

beerose commented:
1. I implemented the solution with an estimated count (show estimate and then change to the exact count), but it feels kinda odd when the number of pages changes. 

2. I added resetting bulk select on page change.

rikinsk commented:
The count changing does feel a little weird. I think I have a solution though (this should be the final change for this PR hopefully). 
Instead of displaying the estimated count and then replacing it with the actual count, lets display a spinner in place of the count till the actual count loads. Lets use the estimated count to only set the estimated number of pages. This way when the count finally loads, the spinner will get replaced by the count (giving an appearance that some new information was loaded) and the number of pages changing will not feel weird.

beerose commented:
Just to confirm, by not displaying the count you mean removing the whole `of X` part like below?

<img width="405" alt="Screenshot 2020-04-04 at 17 13 56" src="https://user-images.githubusercontent.com/9019397/78454327-ad6d9280-7697-11ea-8ae8-d5d2aab44858.png">

Btw, I read the comment before the edit, so I'll leave here how spinner would look like just for a reference.

<img width="300" src="https://user-images.githubusercontent.com/9019397/78454482-82d00980-7698-11ea-9851-123aa36416eb.gif" />

rikinsk commented:
Not exactly. What I meant was this. 

- When count is not loaded (but we have estimated count)

  Count display:
  ![Screenshot from 2020-04-04 20-59-14](https://user-images.githubusercontent.com/54616969/78454738-8921ae00-76b7-11ea-82e3-a26997966e4f.png)

  Pagination: 
  ![Screenshot from 2020-04-04 21-00-05](https://user-images.githubusercontent.com/54616969/78454761-b0787b00-76b7-11ea-9af9-ba5af7948774.png)

- When count loads

  Count display:
  
  ![Screenshot from 2020-04-04 20-58-51](https://user-images.githubusercontent.com/54616969/78454794-e9185480-76b7-11ea-8a46-22296381befe.png)

  Pagination:
  ![Screenshot from 2020-04-04 20-59-25](https://user-images.githubusercontent.com/54616969/78454781-d56cee00-76b7-11ea-83ed-37ef15c0192c.png)

wawhal commented on console/src/components/Services/Data/TableBrowseRows/ViewActions.js:
```diff
@@ -89,13 +86,12 @@ const vMakeRequest = () => {
       data => {
-        const currentTable = getState().tables.currentTable;
-        if (originalTable === currentTable) {
+        if (originalTable === tables.currentTable) {
```
This will always be true?

rikinsk replied:
Nope. currentTable is derived from state again in the fn and can be different as it async because it is in `then()` of requestAction.

rikinsk replied:
Sorry. this seems like another instance of the issue I fixed

wawhal commented on console/src/components/Services/Data/TableBrowseRows/FilterActions.js:
```diff
@@ -108,7 +108,7 @@ const runQuery = tableSchema => {
     }
     dispatch({ type: 'ViewTable/V_SET_QUERY_OPTS', queryStuff: newQuery });
-    dispatch(vMakeRequest());
+    dispatch(vMakeRequest()).then(() => dispatch(vMakeCountRequest()));
```
This is done twice, here and once in `ViewTable` as well.
Do you think there should be a function called `vMakeTableRequests` that abstracts out this part? Something like:

```js
const vMakeTableRequests = () => (dispatch, getState) => {
  dispatch(vMakeRequest()).then(() => dispatch(vMakeCountRequest()));
};
```
You can then call this one function everywhere. Later, if somebody is iterating some part of this code, they might not know to make the change in both places.

rikinsk commented:
/heroku deploy

wawhal commented on console/src/components/Services/Data/TableBrowseRows/ViewActions.js:
```diff
@@ -51,34 +60,22 @@ const vSetDefaults = () => ({ type: V_SET_DEFAULTS });
 const vMakeRequest = () => {
   return (dispatch, getState) => {
-    const state = getState();
+    const { tables } = getState();
```
### Suggestion

I think it would make sense to extract the relevant parts here itself rather than rocking the `tables` object everywhere. This is because, in large schemas, the `tables` object could be **huge**.

You can just:

```js
const {
  currentTable: originalTable,
  currentSchema
} = getState().tables;
```

wawhal commented on console/src/components/Services/Data/TableBrowseRows/ViewActions.js:
```diff
@@ -113,27 +109,50 @@ const vMakeRequest = () => {
   };
 };
 
+const vMakeCountRequest = () => {
```
Same review as the `vMakeRequest` function.

wawhal commented on console/src/components/Common/utils/v1QueryUtils.js:
```diff
@@ -279,3 +279,38 @@ export const resetMetadataQuery = {
+  },
+});
+
+export const getFetchManualTriggersQuery = tableName => ({
```
@beerose @rikinsk This should ideally accept a table definition rather than just a table name. Can we assign this to somebody ASAP? 5 min fix.
